commands/pipelines/diff.js

```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+let cli          = require('heroku-cli-util');
+let co           = require('co');
+let bluebird     = require('bluebird');
+let request      = bluebird.promisify(require('request'));
+let _            = require('lodash');
+
+const PROMOTION_ORDER = ['development', 'staging', 'production'];
+const V3_HEADER = 'application/vnd.heroku+json; version=3';
+const PIPELINES_HEADER = V3_HEADER + '.pipelines';
+
+// Helper functions
+
+function kolkrabbiRequest(url, token) {
+  return request({
+    method: 'GET',
+    url: url,
+    headers: {
+      authorization: 'Bearer ' + token
+    },
+    json: true
+  }).spread(function (res, body) {
+    if (res.statusCode === 404) {
+      let err = new Error('404');
+      err.name = 'NOT_FOUND';
+      throw err;
+    } else if (res.statusCode >= 400) {
+      // TODO: This could potentially catch some 4xx errors that we might want to handle with a
+      // specific error message.
+      throw new Error('failed to fetch diff because of an internal server error.');
+    }
+    return body;
+  });
+}
+
+function* getLatestCommitHash(heroku, appName, appId) {
+  const release = yield heroku.request({
+    method: 'GET',
+    path: `/apps/${appId}/releases`,
+    headers: { 'Accept': V3_HEADER, 'Range': 'version ..; order=desc,max=1' },
+    partial: true
+  });
+  const slug = yield heroku.request({
+    method: 'GET',
+    path: `/apps/${appId}/slugs/${release[0].slug.id}`,
+    headers: { 'Accept': V3_HEADER }
+  });
+  return { name: appName, hash: slug.commit };
+}
+
+function* diff(sourceApp, downstreamApp, repo, githubToken, herokuUserAgent) {
+  if (sourceApp.hash === downstreamApp.hash) {
+    console.log(`\neverything is up to date between ${sourceApp.name} and ${downstreamApp.name}`);
+    return;
+  }
+  const githubDiff = yield request({
+    url: `https://api.github.com/repos/${repo}/compare/${downstreamApp.hash}...${sourceApp.hash}`,
+    headers: {
+      authorization: 'token ' + githubToken,
+      'user-agent': herokuUserAgent,
+    },
+    json: true
+  }).get(1);
+
+  console.log(`\n${sourceApp.name} is ahead of ${downstreamApp.name} by ${githubDiff.ahead_by} commit${githubDiff.ahead_by === 1 ? '' : 's'}:`);
+  for (let i = githubDiff.commits.length - 1; i >= 0; i--) {
```

Comment:
By chance I tested this on a pipeline where the commit sha of the production app is no longer available on GitHub - I think I must've rebased and force pushed or something. Anyway, that case causes an exception on this line:

```sh
TypeError: Cannot read property 'length' of undefined
    at diff (/Users/aappleton/projects/heroku-pipelines/commands/pipelines/diff.js:67:34)
    at next (native)
    at onFulfilled (/Users/aappleton/projects/heroku-pipelines/node_modules/heroku-cli-util/node_modules/co/index.js:65:19)
    at tryCatcher (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/util.js:26:23)
    at Promise._settlePromiseFromHandler (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:507:31)
    at Promise._settlePromiseAt (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:581:18)
    at Promise._settlePromises (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/promise.js:697:14)
    at Async._drainQueue (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:123:16)
    at Async._drainQueues (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:133:10)
    at Immediate.Async.drainQueues [as _onImmediate] (/Users/aappleton/projects/heroku-pipelines/node_modules/bluebird/js/main/async.js:15:14)
    at processImmediate [as _immediateCallback] (timers.js:368:17)
```